### DISCUSSION TITLE
[RFC] ConfigProvider theme 支持传入 algorithm 数组

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#36966] [RFCs - archive]
<a name="Summary"></a>
## Summary
将 ConfigProvider 的 `theme`参数中的 `algorithm` 的类型改为一个数组，支持传入多个 algorithm function 构成 MapToken。
<a name="Motivation"></a>
## Motivation
目前 algorithm 仅支持传入一个算法来从 Seed Token 生成 Map Token，默认提供了亮色和暗色两种算法。但是这两个算法其实仅改了部分 Map Token，还有一些尺寸相关的 Map Token 没有治理。考虑到后面会产出的 紧凑模式 算法，我们可以将亮色和暗色两种模式任一与紧凑模式结合，这就需要 theme 提供接收多种算法来共同构成 Map Token 的能力。同时用户也可以基于某一算法之上继续覆盖，而不用关心上一个算法的内部实现。

<a name="API"></a>
## API
主要是针对 `ThemeConfig`的 `algorithm` 做出改动。
```typescript
type ThemeConfig = {
  token: Partial<SeedToken>,
  algorithm: ((seedToken: SeedToken, mapToken: MapToken) => MapToken)[],
  override: OverrideToken,
  hashed: boolean,
}
```
支持数组，越往后的越晚生效（会覆盖之前的算法结果）。

由于传入的是数组，所以执行上会有先后顺序，所以增加了第二个参数 `mapToken: MapToken`，顺序执行时会将当前处理完的 Map Token 传入，让用户可以对流水线上当前的 Map Token 有感知，并且可以针对性地修改。

同时 antd 将提供一个默认的算法用于帮助用户填充 mapToken。

<a name="9afe3fae"></a>
## Basic Example
```tsx
import { theme } from antd;

const { darkAlgorithm, compactAlgorithm } = theme;

const MyButton = (...props) => {
  return (
    <ConfigProvider theme={{ algorithm: [darkAlgorithm, compactAlgorithm] }}>
      <Button {...props} />
    </ConfigProvider>
  );
}

export default MyButton;
```

<a name="Drawbacks"></a>
## Drawbacks
No obvious drawbacks yet.

**Comment by afc163** [comment]
现在的 compactAlgorithm 代码长啥样的

**Reply by MadCcc** [reply]
还没有，拿出来举个例子的，应该只会影响 margin 之类的 number 类型的 token。

**Comment by MadCcc** [comment]
https://github.com/ant-design/ant-design/blob/next/components/theme/themes/shared/genColorMapToken.ts
目前已经对不同分类的 Map Token 有分开治理，也是为了最终形态铺路。

**Comment by zombieJ** [comment]
算法这个感觉还可以再提一下，如果只返回 MapToken 说明是全部，反之可以返回低优先级：
```tsx
type Algorithm = (seed: SeedToken) => (MapToken | { lowPriority: MapToken, highPriority: MapToken }) 
```

比如 Dark 返回的就是 MapToken，而 compact 返回的则是 `{ lowPriority, highPriority }`，当前面存在算法时，后面的算法仅仅会覆盖 highPriority。

**Comment by zombieJ** [comment]
或者，干脆算法还能再接受额外的 MapToken：
```tsx
type Algorithm = (seed: SeedToken, pipeToken?: AliasToken) => MapToken
```

如果你的算法想在之前算法上继续改动，那就可以处理 pipeToken，反之不处理就是覆盖。

**Reply by MadCcc** [reply]
这个可以，给用户自由度，但是不限制用户
